Lib/site.py

```diff
@@ -558,6 +575,12 @@ def main():
     setquit()
     setcopyright()
     sethelper()
+    # Detect REPL by prompt, -i flag or PYTHONINSPECT env var being set.
+    if (
+        hasattr(sys, 'ps1') or sys.flags.interactive or
+        os.environ.get('PYTHONINSPECT')
+    ):
+        _register_detect_pip_usage_in_repl()
```

Comment:
I'd suggest a structural adjustment here:

1. Factor this out to a `_set_interactive_excepthook` helper function
2. Call that function after `execsitecustomize` rather than before (so the customization modules are run with only the original hook installed)

Reply:
I think I might be missing your full intent on point 1 here Nick. Are you thinking of a new function like this?

```python
def _set_interactive_excepthook():
    # Detect REPL by prompt, -i flag or PYTHONINSPECT env var being set.
    if (hasattr(sys, 'ps1') or sys.flags.interactive or
        os.environ.get('PYTHONINSPECT')):
        _register_detect_pip_usage_in_repl()
```

Reply:
This thread is superseded by the [discussion below](https://github.com/python/cpython/pull/8536#discussion_r207061225):

> I've refactored a bit with a new `_set_interactive_hook` function.